File: api/analyst-ratings.py
"""
BIST analist tavsiye + hedef fiyat — toplu borsapy serverless endpoint.
/radar için sembol başına hedef fiyat, yükseliş potansiyeli ve AL/TUT/SAT
dağılımını tek istekte döner (per-sembol /api/bist-fundamentals çok pahalı).

URL: /api/analyst-ratings?symbols=THYAO,ASELS
Cache: 6 saat (analist verisi yavaş değişir — Vercel edge cache).
"""

from http.server import BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs
import json
import math
import traceback

MAX_SYMBOLS = 60  # tek istekte üst sınır — TS tarafı evreni chunk'lar


def num(x):
    try:
        if x is None:
            return None
        v = float(x)
        return None if (math.isnan(v) or math.isinf(v)) else v
    except (TypeError, ValueError):
        return None


def fetch_one(bp, symbol):
    """Tek sembol için analist tavsiye + hedef fiyat + dağılım.
    Her alt-bölüm kendi try/except'inde — biri düşerse diğerleri gelir."""
    t = bp.Ticker(symbol)
    out = {"symbol": symbol}

    try:
        rec = t.recommendations
        if isinstance(rec, dict):
            out["recommendation"] = rec.get("recommendation")
            out["target_price"] = num(rec.get("target_price"))
            out["upside_pct"] = num(rec.get("upside_potential"))
    except Exception as e:
        print(f"[analyst-ratings] {symbol} recommendations error: {e}")

    try:
        tg = t.analyst_price_targets
        if isinstance(tg, dict):
            out["target_mean"] = num(tg.get("mean"))
            out["num_analysts"] = tg.get("numberOfAnalysts")
    except Exception as e:
        print(f"[analyst-ratings] {symbol} price_targets error: {e}")

    try:
        s = t.recommendations_summary
        if isinstance(s, dict):
            out["strong_buy"] = s.get("strongBuy")
            out["buy"] = s.get("buy")
            out["hold"] = s.get("hold")
            out["sell"] = s.get("sell")
            out["strong_sell"] = s.get("strongSell")
    except Exception as e:
        print(f"[analyst-ratings] {symbol} summary error: {e}")

    # target_mean yoksa recommendations.target_price'a düş
    if out.get("target_mean") is None and out.get("target_price") is not None:
        out["target_mean"] = out["target_price"]
    return out


class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        out = {"status": "ok", "ratings": {}}
        status_code = 200
        try:
            qs = parse_qs(urlparse(self.path).query)
            symbols = [
                s.strip().upper()
                for s in qs.get("symbols", [""])[0].split(",")
                if s.strip()
            ][:MAX_SYMBOLS]

            import borsapy as bp

            for sym in symbols:
                try:
                    out["ratings"][sym] = fetch_one(bp, sym)
                except Exception as e:
                    print(f"[analyst-ratings] {sym} error: {e}")
        except ImportError as e:
            out = {"status": "error", "message": f"borsapy import error: {e}"}
            status_code = 500
        except Exception as e:
            out = {"status": "error", "message": str(e), "trace": traceback.format_exc()[:2000]}
            status_code = 500

        body = json.dumps(out, ensure_ascii=False, default=str).encode("utf-8")
        self.send_response(status_code)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Cache-Control", "public, max-age=0, s-maxage=21600, stale-while-revalidate=43200")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

File: src/app/(app)/_lib/analyst-ratings.ts
"use server";

// BIST analist tavsiye + hedef fiyat — /api/analyst-ratings (borsapy) toplu
// endpoint'i. /radar için sembol başına hedef fiyat, yükseliş potansiyeli ve
// AL/TUT/SAT dağılımı. Per-sembol /api/bist-fundamentals çok pahalı olduğundan
// bu toplu endpoint chunk'lı çekilir (bist-history ile aynı desen).

export interface AnalystRating {
  symbol: string;
  recommendation: string | null;
  target_mean: number | null;
  upside_pct: number | null;
  strong_buy: number | null;
  buy: number | null;
  hold: number | null;
  sell: number | null;
  strong_sell: number | null;
  num_analysts: number | null;
}

interface RawRating {
  symbol: string;
  recommendation?: string | null;
  target_price?: number | null;
  target_mean?: number | null;
  upside_pct?: number | null;
  strong_buy?: number | null;
  buy?: number | null;
  hold?: number | null;
  sell?: number | null;
  strong_sell?: number | null;
  num_analysts?: number | null;
}

function host(): string {
  const prodHost = process.env.VERCEL_PROJECT_PRODUCTION_URL || process.env.VERCEL_URL;
  return prodHost ? `https://${prodHost}` : "http://localhost:3000";
}

const toNum = (v: unknown): number | null =>
  typeof v === "number" && Number.isFinite(v) ? v : null;

async function fetchBatch(symbols: string[]): Promise<Record<string, RawRating>> {
  if (symbols.length === 0) return {};
  try {
    const res = await fetch(
      `${host()}/api/analyst-ratings?symbols=${encodeURIComponent(symbols.join(","))}`,
      { next: { revalidate: 21600, tags: ["analyst-ratings"] } },
    );
    if (!res.ok) return {};
    const json = (await res.json()) as { status: string; ratings?: Record<string, RawRating> };
    if (json.status !== "ok" || !json.ratings) return {};
    return json.ratings;
  } catch (err) {
    console.error("[analyst-ratings] batch error", err);
    return {};
  }
}

/** BIST sembolleri için analist tavsiye/hedef fiyat — 20'lik batch'ler paralel. */
export async function getAnalystRatings(
  symbols: string[],
): Promise<Record<string, AnalystRating>> {
  if (symbols.length === 0) return {};
  const uniq = Array.from(new Set(symbols));
  const BATCH = 20;
  const batches: string[][] = [];
  for (let i = 0; i < uniq.length; i += BATCH) batches.push(uniq.slice(i, i + BATCH));
  const results = await Promise.all(batches.map((b) => fetchBatch(b)));

  const out: Record<string, AnalystRating> = {};
  for (const chunk of results) {
    for (const [sym, r] of Object.entries(chunk)) {
      out[sym] = {
        symbol: sym,
        recommendation: r.recommendation ?? null,
        target_mean: toNum(r.target_mean ?? r.target_price),
        upside_pct: toNum(r.upside_pct),
        strong_buy: toNum(r.strong_buy),
        buy: toNum(r.buy),
        hold: toNum(r.hold),
        sell: toNum(r.sell),
        strong_sell: toNum(r.strong_sell),
        num_analysts: toNum(r.num_analysts),
      };
    }
  }
  return out;
}
